pkg/rbd/controllerserver.go

```diff
@@ -95,7 +95,7 @@ func (cs *ControllerServer) parseVolCreateRequest(ctx context.Context, req *csi.
 	}
 
 	// if it's NOT SINGLE_NODE_WRITER and it's BLOCK we'll set the parameter to ignore the in-use checks
-	rbdVol, err := genVolFromVolumeOptions(ctx, req.GetParameters(), nil, (isMultiNode && isBlock), false)
+	rbdVol, err := genVolFromVolumeOptions(ctx, req.GetParameters(), req.GetSecrets(), (isMultiNode && isBlock), false)
```

Comment:
I think we should init KMS - it's not much overhead on the process, but this way we'll verify that KMS is configured correctly (fail fast).

If we skip KMS initialisation for provisioner that means we'll create an image in ceph, but when we try to attach it - it will fail due to misconfiguration... I believe failing fast is better user experience.

Reply:
fail-fast is fine. I would possibly on CSI init check if the KMS config is correct, rather than at each CreateVolume call though to fail fast. There is the outlier that an new KMS ID was added later, during runtime, but barring those we should catch config issues once during init if possible.

There are various scenarios where speed of volume creation is crucial, so optimizing those would be helpful.

Reply:
I don't think it is possible to validate the KMS config on CSI init:

the KMS type is defined in storageclass, so we can't just assume that all entries from kms config are vault configurations (currently this is the only supported type but the interface is generic enough that other KMS would be easy to add - i.e. k8s secrets to store passphrase per volume in secrets).

In order to optimise scenario you've mentioned - we can introduce another flag `validateKMSOnCreate: true` that can be used to bypass the KMS validation and instantiation

Reply:
> In order to optimise scenario you've mentioned - we can introduce another flag `validateKMSOnCreate: true` that can be used to bypass the KMS validation and instantiation

No new flags please :) (I am config options blow up averse) let's hold onto this for now, as KMS initialization happens only when encryption is turned on, so we can evaluate and later optimize if required.